server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -347,4 +357,113 @@ public String toString() {
             }
         }
     }
+
+    private class ConnectionHistory {
+        record NodeConnectionHistory(String ephemeralId, long disconnectTime, Exception disconnectCause) {}
+
+        /**
+         * Holds the DiscoveryNode nodeId to connection history record.
+         *
+         * Entries for each node are reserved during NodeConnectionsService.connectToNodes, by placing a (nodeId, dummy) entry
+         * for each node in the cluster. On node disconnect, this entry is updated with its NodeConnectionHistory. On node
+         * connect, this entry is reset to the dummy value. On NodeConnectionsService.disconnectFromNodesExcept, node entries
+         * are removed.
+         *
+         * Each node in the cluster always has a nodeHistory entry that is either the dummy value or a connection history record. This
+         * allows node disconnect callbacks to discard their entry if the disconnect occurred because of a change in cluster state.
+         */
+        private final NodeConnectionHistory dummy = new NodeConnectionHistory("", 0, null);
+        private final ConcurrentMap<String, NodeConnectionHistory> nodeHistory = ConcurrentCollections.newConcurrentMap();
+
+        ConnectionHistory() {
+            NodeConnectionsService.this.transportService.addConnectionListener(new TransportConnectionListener() {
+                @Override
+                public void onNodeConnected(DiscoveryNode node, Transport.Connection connection) {
+                    // log case where the remote node has same ephemeralId as its previous connection
+                    // (the network was disrupted, but not the remote process)
+                    NodeConnectionHistory nodeConnectionHistory = nodeHistory.get(node.getId());
+                    if (nodeConnectionHistory != null) {
+                        nodeHistory.replace(node.getId(), nodeConnectionHistory, dummy);
+                    }
```

Comment:
This looks a little racy, although in practice I think it's fine because `ClusterConnectionManager` protects against opening multiple connections to the same node concurrently. Still, if we did all this (including the logging) within a `nodeHistory.compute(node.getId, ...)` then there'd obviously be no races.